### ISSUE TITLE
CSS naming convention

### ISSUE THREAD

**Issue by redaxmedia** [uikit/uikit#2820]
Dear UIKit team,

this is not a bug report - just an suggestion for a CSS naming convention.

Prolog
--------

I never was a friend of BEM and therefore developed NCSS a couple of years ago. It seems that the UIKit team is accidental following many of the rules and conventions like NCSS does.

Further reading: https://ncss.io/documentation/classes || https://ncss.io/documentation/examples

Goal
------

A predictable grammar for CSS that provides semantic information about the HTML template.

- What tags, components and sections are affected
- What is the relation of one class to another

> You understand the structure of the website while reading the CSS

Linter
-------

There is an NodeJS powered linter that validates HTML against the NCSS naming convention. I hacked the original ruleset to match the UIKit convention. Of course, this is cheating the idea behind NCSS :-)

Further reading: https://ncss.io/documentation/basics/linter-1002

```
bin/cmd.js --url=https://getuikit.com/docs/table --namespace=uk --loglevel=info --selector='*:not([class*=hljs])'
```

![grafik](https://user-images.githubusercontent.com/1835397/28934521-8add0942-7881-11e7-8b66-fc1ac973f2f5.png)

Result
--------

There are couple of missing and mixed namespaces:

```
FAILED! (36 errors found)

error: invalid namespace > ul.uk-nav.uk-nav-default.tm-nav
error: invalid namespace > h1.uk-h1.tm-heading-fragment
error: invalid namespace > h2.uk-h3.tm-heading-fragment
error: invalid namespace > a.js-copy
error: invalid namespace > a.js-codepen
error: invalid namespace > h2.uk-h3.tm-heading-fragment
error: invalid namespace > a.js-copy
error: invalid namespace > a.js-codepen
error: invalid namespace > h2.uk-h3.tm-heading-fragment
error: invalid namespace > a.js-copy
error: invalid namespace > a.js-codepen
error: invalid namespace > h2.uk-h3.tm-heading-fragment
error: invalid namespace > a.js-copy
error: invalid namespace > a.js-codepen
error: invalid namespace > h2.uk-h3.tm-heading-fragment
error: invalid namespace > a.js-copy
error: invalid namespace > a.js-codepen
error: invalid namespace > h2.uk-h3.tm-heading-fragment
error: invalid namespace > a.js-copy
error: invalid namespace > a.js-codepen
error: invalid namespace > h2.uk-h3.tm-heading-fragment
error: invalid namespace > a.js-copy
error: invalid namespace > a.js-codepen
error: invalid namespace > h2.uk-h3.tm-heading-fragment
error: invalid namespace > a.js-copy
error: invalid namespace > a.js-codepen
error: invalid namespace > h3.uk-h4.tm-heading-fragment
error: invalid namespace > a.js-copy
error: invalid namespace > a.js-codepen
error: invalid namespace > h2.uk-h3.tm-heading-fragment
error: invalid namespace > a.js-copy
error: invalid namespace > a.js-codepen
error: invalid namespace > h2.uk-h3.tm-heading-fragment
error: invalid namespace > a.js-copy
error: invalid namespace > a.js-codepen
error: invalid namespace > ul.uk-nav.uk-nav-default.tm-nav
```

Related ruleset: https://pastebin.com/dJ5uz6jg

tl;dr
-----

UIKit is accidental following many of the rules and conventions of NCSS. I suggest to refactor UIKit to NCSS or use a customized ruleset for quality assurance via NCSS Linter and CircleCI.

Further reading: https://ncss.io

**Comment by saschadube** [comment]
Hi, thanks for the feedback 😄 

Why would we need to refactor UIkit? The classes are named fine. The error you found, are the theme classes. They have a different name space ```tm-```

What do you suggest?

**Comment by redaxmedia** [comment]
There is no need to refactor UIkit as long you are fine not to follow a naming convention. I know this is not a good moment anyway because you are in beta state and want to release asap.

I opened this issue because the naming inside UIKit is in fact very similar to NCSS accept the fact of a predictable grammar based on a ruleset that can be validated with a linter.

At this point, I suggest to start a discussion in your team. Otherwise feel free to close this issue.
